### ISSUE TITLE
Crash at Windows Shutdown

### ISSUE THREAD

**Issue by Y4m4k** [Artemis-RGB/Artemis#632]
Hi - love the program… learning it pretty quick. I have an issue when Artemis is running in the background and I shut down the computer (running Windows 10) Artemis crashes and doesn’t allow the computer to shut down (image attached). It happens every time the Windows 10 shut down sequence begins. Artemis is linked to OpenRGB to acquire the LEDs off the OpenRGB SDK server.

Let me know what other info you need.

Thanks for the help!

![73F804CE-BB2E-45AA-A036-62BA2098FD9A](https://user-images.githubusercontent.com/86619370/123672348-34fb7000-d80d-11eb-8ca6-7639d29ff0ec.jpeg)

**Comment by SpoinkyNL** [comment]
Hey there, 

I've seen this one before but we couldn't find a solution then. DDE Server Window is some kind of old technology used by Excel, Artemis doesn't use it at all.

So I'm a bit clueless about this one, can you check your Windows Event Viewer and see if you can find  any Artemis-related errors there?

**Comment by Cheerpipe** [comment]
You could also disable plugins one by one to see if this is Artemis or plugin related.

**Comment by Y4m4k** [comment]
Thanks both. 

@SpoinkyNL I checked the Windows Event Viewer, it didn't give more information than the first picture I added.

@Cheerpipe I did as you recommended and disabled unnecessary plugins. I kept the bare bones but I still get the error when I shut down. This time it doesn't say DDE Server Window, however, same error code but location is different. Do you want me to share the Artemis log? Do you have any recommendations on what to do. Do I need OpenRGB to detect the RGB natively? That what I could disable the OpenRGB plugin.

Thanks for the help!

![IMG_1602](https://user-images.githubusercontent.com/86619370/123685710-fa99cf00-d81c-11eb-92f0-4bf195521af9.jpg)

**Comment by SpoinkyNL** [comment]
Hmm strange, I'd expect you to find something in vent Viewer (Local) > Windows Logs > Application

**Comment by Cheerpipe** [comment]
What  if you continue disabling plugins until there is no plugin enabled?

O know this may be a lot of effort by may you disable plugins 1 by one and restart and see what happens?

**Comment by Y4m4k** [comment]
Hey @SpoinkyNL

I found the Artemis error in the Windows Event Viewer. Have a look, let me know if you need me to do anything. I had installed Microsoft Windows Desktop Runtime .NET - 5.0.7(x64) before installing Artemis.

@Cheerpipe - if this doesn't work, I'll give that a shot.

Thanks for the help!

**Comment by Y4m4k** [comment]
Sorry forgot to add the image of the error.

@SpoinkyNL @Cheerpipe 

![Artemis Event Log](https://user-images.githubusercontent.com/86619370/123688788-721d2d80-d820-11eb-9b3f-11b6f2e2fba3.PNG)

**Comment by SpoinkyNL** [comment]
There's the money shot, thank you :)
I'll update this issue once I've released a fix

**Comment by Y4m4k** [comment]
@SpoinkyNL awesome, looking forward to the update :-D !! Anything I can do in the meantime to shut down normally?

@Cheerpipe - don't know what the even means, sounds like a Catch 22 to me!!

**Comment by SpoinkyNL** [comment]
This contains the fix, if you extract this into the Artemis install directory (default `C:\Program Files\Artemis`) it should be fixed.
Next update the file will be overwritten again, but then it should include the fix as well.

[Artemis.UI.zip](https://github.com/Artemis-RGB/Artemis/files/6728992/Artemis.UI.zip)

**Comment by Y4m4k** [comment]
@SpoinkyNL thank you! I extracted and overwrote the previous data in C:\Program Files\Artemis. Windows shutdown normal with no errors 😁. Glad you were able to fix it so quickly 👍

Just out of curiosity,  what was causing the issue?

**Comment by SpoinkyNL** [comment]
Great :)

The issue is that during Windows shutdown Windows asks all applications to gracefully close.

Artemis also does this, getting rid of all its internals, closing the connection with OpenRGB etc. etc. 

One of these things during shutdown is causing a non fatal exception and Artemis is 'catching' that exception to avoid the application crashing altogether. Whenever this happens the error is logged and nothing else happens. However, because Artemis was already in the progress of shutting down for Windows, writing to the log failed and threw _another_ exception. That exception however happened in the final 'catch' so there was nowhere left for the exception to go, resulting in a crash ;p

**Comment by Y4m4k** [comment]
Hey, thanks for explaining it in layman terms! I always had a curiosity for coding, but could never catch on, maybe next life time 😊

Take care and thanks for the help and extremely fast turn around!

**Comment by overdarki** [comment]
Hello, i still get some crashes on shutdown every time, maybe that is the same issue.

**Artemis**: 20210711.1
**Win10**: 21H1

**Error**:
Artemis.Core.ArtemisCoreException: Cannot disable this plugin because it still has enabled features
   at Artemis.Core.Plugin.SetEnabled(Boolean enable) in D:\a\1\s\Artemis\src\Artemis.Core\Plugins\Plugin.cs:line 289
   at Artemis.Core.Plugin.Dispose(Boolean disposing) in D:\a\1\s\Artemis\src\Artemis.Core\Plugins\Plugin.cs:line 213
   at Artemis.Core.Plugin.Dispose() in D:\a\1\s\Artemis\src\Artemis.Core\Plugins\Plugin.cs:line 313
   at Artemis.Core.Services.PluginManagementService.UnloadPlugin(Plugin plugin) in D:\a\1\s\Artemis\src\Artemis.Core\Services\PluginManagementService.cs:line 465
   at Artemis.Core.Services.PluginManagementService.UnloadPlugins() in D:\a\1\s\Artemis\src\Artemis.Core\Services\PluginManagementService.cs:line 280
   at Artemis.Core.Services.PluginManagementService.Dispose() in D:\a\1\s\Artemis\src\Artemis.Core\Services\PluginManagementService.cs:line 199
   at Artemis.Core.Services.CoreService.Dispose() in D:\a\1\s\Artemis\src\Artemis.Core\Services\CoreService.cs:line 186
   at Ninject.Activation.Strategies.DisposableStrategy.<>c.<Deactivate>b__0_0(IDisposable x)
   at Ninject.Activation.Strategies.DisposableStrategy.Deactivate(IContext context, InstanceReference reference)
   at Ninject.Activation.Pipeline.<>c__DisplayClass7_0.<Deactivate>b__0(IActivationStrategy s)
   at Ninject.Infrastructure.Language.ExtensionsForIEnumerableOfT.Map[T](IEnumerable`1 series, Action`1 action)
   at Ninject.Activation.Pipeline.Deactivate(IContext context, InstanceReference reference)
   at Ninject.Activation.Caching.Cache.Forget(CacheEntry entry)
   at Ninject.Activation.Caching.Cache.Forget(IEnumerable`1 cacheEntries)
   at Ninject.Activation.Caching.Cache.Clear()
   at Ninject.KernelBase.Dispose(Boolean disposing)
   at Artemis.UI.ApplicationStateManager.<>c__DisplayClass1_0.<.ctor>b__0(Object _, SessionEndingCancelEventArgs _) in D:\a\1\s\Artemis\src\Artemis.UI\ApplicationStateManager.cs:line 33
   at System.Windows.Application.OnSessionEnding(SessionEndingCancelEventArgs e)
   at System.Windows.Application.WmQueryEndSession(IntPtr lParam, IntPtr& refInt)
   at System.Windows.Application.AppFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
